### PR TITLE
[Governance] vote with connected wallet

### PR DESCRIPTION
**Pull request by zihan-aptos**
https://user-images.githubusercontent.com/109111707/183007941-32e44042-e840-4fde-94b8-deed4d2c0831.mov

### CODE REVIEW

File: src/pages/Governance/Proposal/card/VoteButtons.tsx [hunk 1]
```diff
@@ -25,37 +38,98 @@ export default function VoteButtons({proposalId}: Props) {
...
+          variant="primary"
+          sx={{justifyContent: "start", backgroundColor: negativeColor}}
+          startIcon={<CancelOutlinedIcon />}
+          onClick={() => onVote(false)}
+        >
+          {voteAgainst}
+        </Button>
+      </Stack>
+      <Snackbar
+        open={voteSucceeded === true}
```
[the remark this answers is not in the document]
Comment (zihan-aptos) [reply src/pages/Governance/Proposal/card/VoteButtons.tsx:116]: I thought about enum and got lazy lol. Changing it.

File: src/pages/Governance/hooks/useSubmitVote.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    console.log(response);
+    if ("hash" in response) {
+      return {
+        succeeded: true,
+        transactionHash: response["hash"],
+      };
+    }
+  } catch (error: any) {
+    if (typeof error == "object" && "message" in error) {
+      responseOnFailure.message = error.message;
```
[the remark this answers is not in the document]
Comment (zihan-aptos) [reply src/pages/Governance/hooks/useSubmitVote.ts:37]: The error message we're getting is `{message: "...", code: -30000}`. I feel that `code` doesn't mean anything to the users, `message` should be sufficient. We can revisit this part later. I think first step would be work with the team to return a better message/code (why this txn fails).

File: src/pages/Governance/hooks/useSubmitVote.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    succeeded: false,
+    message: "Unknown Error",
+  };
+
+  try {
+    const response = await window.aptos.signAndSubmitTransaction(
+      transactionPayload,
+    );
+    console.log(response);
+    if ("hash" in response) {
```
[the remark this answers is not in the document]
Comment (zihan-aptos) [reply src/pages/Governance/hooks/useSubmitVote.ts:29]: Right now I don't think we have this scenario. If a vote succeeded, there must be a transaction hash. Again I think we want to work with the team to return a better response/error from the move module and revisit this part.

File: src/pages/Governance/hooks/useSubmitVote.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  transactionPayload: Types.TransactionPayload,
+): Promise<VoteResponse> => {
+  const responseOnFailure = {
+    succeeded: false,
+    message: "Unknown Error",
+  };
+
+  try {
+    const response = await window.aptos.signAndSubmitTransaction(
+      transactionPayload,
```
Comment (msavir) [inline src/pages/Governance/hooks/useSubmitVote.ts:26]: I feel this a wallet api and might be better to place it in the wallet api file...?

File: src/pages/Governance/hooks/useSubmitVote.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    }
+  }
+  return responseOnFailure;
+};
+
+// TODO: confirmation modal
+const useSubmitVote = (proposalId: string) => {
+  const [voteResponse, setVoteResponse] = useState<VoteResponse | null>(null);
+
+  async function submitVote(shouldPass: boolean, ownerAccountAddr: string) {
```
Comment (msavir) [inline src/pages/Governance/hooks/useSubmitVote.ts:46]: not a blocker can revisit - I believe we are going toward BCS transaction
https://github.com/aptos-labs/wallet/pull/85#discussion_r938440809

**Review by msavir — APPROVED** [review]

File: src/pages/Governance/hooks/useSubmitVote.ts (hunk 4 above)
Comment (zihan-aptos) [reply src/pages/Governance/hooks/useSubmitVote.ts:26]: Do you mean the `signAndSubmitTransaction` part should go to a wallet api file? Yeah agree. signing transactions is a rather general functionality and not just voting will need it. Let me add a TODO here. We can change it along with using BCS as you mentioned below.

Comment (msavir) [reply src/pages/Governance/hooks/useSubmitVote.ts:29]: sounds good. I'll be working on the stake transaction so can implement those changes a part of that PR

**Pull request merged**